README.md
# Dnd

---

[![Build Status](https://secure.travis-ci.org/aralejs/dnd.png)](https://travis-ci.org/aralejs/dnd)
[![Coverage Status](https://coveralls.io/repos/aralejs/dnd/badge.png?branch=master)](https://coveralls.io/r/aralejs/dnd)

Drap & Drop

## 配置说明

### element    element[Array]|selector
可拖放的元素, 即源节点
>注: 只取第一个元素, 非法时抛出异常, config的参数非法时将自动取默认值

### containment    element[Array]|selector|null
拖放的边界, 默认null为无边界 
>注: 只取第一个元素, element初始必须要位于containment的内部, 否则将默认为null

### proxy    element[Array]|selector|null
代理元素, 实际上跟随鼠标移动的元素, 默认null为源节点element的clone
>注: 只取第一个元素

### drop    element[Array]|selector|null
放置元素, 默认null为无
>注: 可以有多个元素

### axis   'x'|'y'|false
拖放指定的方向, 默认false为任意方向

### visible    boolean
被拖放的元素在源位置上是否可见, 默认false为不可见

### revert    boolean
是否返回源节点初始位置, 默认false为不返回
>注: 当drop不为null并且没有在drop元素上释放鼠标 或者 按下esc取消, 将返回源节点初始位置, 与revert无关

### revertDuration    number
返回速度, 默认为500
>注: 源节点显示(visible = true)时, 拖放结束时移动到拖放位置的速度也取此值

### disabled    boolean
是否禁止该元素拖放, 默认false为不禁止

### dragCursor    string
拖放过程中没进入放置元素drop时光标形状, 默认为"move"

### dropCursor    string
拖放过程中进入放置元素drop时光标的形状, 默认为"copy"

### zIndex number
代理元素proxy拖放过程中的z-index,  默认为9999


## API

### Dnd(element,  config)
构造函数,  element不能为空
>注: element为空会抛出异常,  config应为纯粹对象

### set(option,  value)
设置配置属性,  element不能设置
>注: 一切set设置应在拖放前

### get(option)
获取配置属性
>注: 配置属性为DOM元素的, 经过拖放后均返回其jquery对象,  proxy将返回源节点clone元素的jquery对象(proxy设置为null时)

### open()
开启页面的拖放功能
>注: 默认use dnd组件就自动open了

### close()
关闭页面的拖放功能


## 事件

### dragstart  (dataTransfer,  dragging,  dropping)
dataTransfer为拖放数据,  dragging为代理元素(元素对象均为jquery对象); 拖放开始时触发（按下鼠标并且至少移动1px),  常用来设置拖放数据dataTransfer
>注: dropping为空, 此处为保持参数一致; 
	
### drag (dragging,  dropping)
dragging为代理元素,  dropping为目标元素
拖放中一直触发, 直到鼠标释放
>注: dropping有可能为空

### dragenter (dragging,  dropping)
dragging为代理元素,  dropping为目标元素
鼠标刚进入drop元素中触发

### dragover (dragging,  dropping)
dragging为代理元素,  dropping为目标元素
鼠标在drop元素中移动一直触发

### dragleave (dragging,  dropping)
dragging为代理元素,  dropping为目标元素
鼠标刚离开drop元素中触发
>注: dropping为离开的drop元素

### drop (dataTransfer,  dragging,  dropping)
dragging为源节点元素,  dropping为放置元素
鼠标在drop元素中释放时触发,  常用来读取dataTransfer值

### dragend (dragging,  dropping)
dragging为源节点元素,  dropping为放置元素; 
拖放结束后触发,  常和dragleave处理相同,  用来取消dragenter中的设置
>注: 当没触发drop时, dropping为null; 按esc时会取消拖放回到源节点初始位置, 但仍会触发dragend


## data-attr实现拖放

###在data-attr上进行配置
data-dnd=true data-config为JSON字符串,  详细见演示
>注: 这种方式不支持dataTransfer和一系列事件,  只是简单拖放


## 最佳实践

带有语义(拖拽数据)的拖放
```javascript
seajs.use(['dnd', '$'], function(Dnd, $){

    var proxy = document.createElement('img'),
        dnd = null ;

    $(proxy).on('load', function(){
        dnd = new Dnd('#drag5', {drop: '#drop2', proxy: proxy, visible: true, 
              revert: true}) ;

        // dataTransfer为拖放数据，传输信息
        dnd.on('dragstart', function(dataTransfer, dragging, dropping){
            dataTransfer.data = '玉伯也叫射雕' ;
        })
        dnd.on('dragenter', function(dragging, dropping){
            dropping.addClass('over') ;
        })
        dnd.on('dragleave', function(dragging, dropping){
            dropping.removeClass('over') ;
        })
        dnd.on('drop', function(dataTransfer, dragging, dropping){
            if(typeof(dataTransfer.data) !== 'undefined'){
                dropping.text(dataTransfer.data) ;
            }
        })
        dnd.on('dragend', function(dragging, dropping){
            if(dropping) dropping.removeClass('over') ;
        })
    })
    $(proxy).css('width', 50) ;
    $(proxy).css('height', 50) ;
    proxy.src = 'http://tp3.sinaimg.cn/1748374882/180/40020642911/1' ;
});
```

用data-attr来实现的简单拖放
```html
<div id="drag6" class="drag" data-dnd=true 
data-config='{"drop": "#drop3", "zIndex": 99}'></div>
<br />
<div id="drop3" class="container"></div>
```













































HISTORY.md
# HISTORY

---

## 1.0.0 `2013-08-27`

`tag:new` 第一个正式版本

tests/dnd-spec.js
define(function(require){

  var $ = require('$'),
      Dnd = require('dnd'),
      expect = require('expect') ;


  describe('Dnd', function(){
      var template = '<style>' + 
                     '.container{width:500px; height:200px;background:#CCC;}' +
                     '.drag{width:50px; height:50px;background:#07B1EE;}' +
                     '</style>' +
                     '<div id="wrap"><div id="container1" class="container">' +
                     '<div id="drag1" class="drag"></div></div><br/>' +
                     '<div id="drop1" class="container"></div></div>', 
          element = null,
          dnd = null ;
      
      beforeEach(function(){
          element = $(template).appendTo(document.body);
      });
      
      afterEach(function(){
          element.remove() ;
          dnd = null ;
          element = null ;
      });
      
      
      
      /*
       * 测试用例比较位置数值时，均加了parseInt
       * 主要由于IE高版本浏览器(10 & 11)对offset返回值的差异
      */
      
      
      // 测试构造
      it('instance', function(done){
          // 非法构造
          try{
              dnd = new Dnd() ;
          } catch(error){
              expect(error.message).to.be('element error!') ;
          }
          
          // 一般构造
          dnd = new Dnd('#drag1') ;
          expect(dnd.get('element').get(0)).to.be($('#drag1').get(0)) ;
          expect(dnd.get('containment')).to.be(null) ;
          expect(dnd.get('axis')).to.be(false) ;
          expect(dnd.get('visible')).to.be(false) ;
          expect(dnd.get('proxy')).to.be(null) ;
          expect(dnd.get('drop')).to.be(null) ;
          expect(dnd.get('revert')).to.be(false) ;
          expect(dnd.get('revertDuration')).to.be(500) ;
          expect(dnd.get('disabled')).to.be(false) ;
          expect(dnd.get('dragCursor')).to.be('move') ;
          expect(dnd.get('dropCursor')).to.be('copy') ;
          expect(dnd.get('zIndex')).to.be(9999) ;
          
          // 不合法配置mousedown后, 自动转化为默认配置
          dnd = new Dnd('#drag1', {containment: '#drop1', proxy: '#drag1',
                drop: '#drag1', axis: 0, visible: 0, revert: 0,      
                revertDuration:'aaa', disabled:0, dragCursor:0, dropCursor:0, 
                zIndex: 'aaa'}) ;
          expect(dnd.get('containment')).to.be('#drop1') ;
          expect(dnd.get('proxy')).to.be('#drag1') ;
          expect(dnd.get('drop')).to.be('#drag1') ;
          expect(dnd.get('axis')).to.be(0) ;
          expect(dnd.get('visible')).to.be(0) ;
          expect(dnd.get('revert')).to.be(0) ;
          expect(dnd.get('revertDuration')).to.be('aaa') ;
          expect(dnd.get('disabled')).to.be(0) ;
          expect(dnd.get('dragCursor')).to.be(0) ;
          expect(dnd.get('dropCursor')).to.be(0) ;
          expect(dnd.get('zIndex')).to.be('aaa') ;
          $('#drag1').on('mousedown', function(event, x, y){
              event.which = 1 ;
          }) ;
          $('#drag1').trigger('mousedown') ;
          $(document).trigger('mouseup') ;
          setTimeout(function(){
              expect(dnd.get('containment')).to.be(null) ;
              expect(dnd.get('proxy').get(0) !==   
                      $('#drag1').get(0)).to.be(true) ;
              expect(dnd.get('drop')).to.be(null) ;
              expect(dnd.get('axis')).to.be(false) ;
              expect(dnd.get('visible')).to.be(false) ;
              expect(dnd.get('revert')).to.be(false) ;
              expect(dnd.get('revertDuration')).to.be(500) ;
              expect(dnd.get('disabled')).to.be(false) ;
              expect(dnd.get('dragCursor')).to.be('move') ;
              expect(dnd.get('dropCursor')).to.be('copy') ;
              expect(dnd.get('zIndex')).to.be(9999) ;
              done() ;
          }, 80) ;
      });
      
      
      
      // 测试两阶段拖放
      it('drag', function(done){
          var originx = $('#drag1').offset().left ;
              originy = $('#drag1').offset().top ;
              lastx = 0,
              lastx = 0 ;
          
          dnd = new Dnd('#drag1') ;
          $('#drag1').on('mousedown', function(event, x, y){
              event.which = 1 ;
              lastx = event.pageX = originx + x;
              lasty = event.pageY = originy + y ;
          }) ;
          $('#drag1').on('mousemove', function(event, x, y){
              lastx = event.pageX = lastx + x ;
              lasty = event.pageY = lasty + y ;
          }) ;
          
          $('#drag1').trigger('mousedown', [20, 20]) ;
          $('#drag1').trigger('mousemove', [200, 300]) ;
          $(document).trigger('mouseup') ;
          $('#drag1').trigger('mousedown', [20, 20]) ;
          $('#drag1').trigger('mousemove', [200, 300]) ;
          $(document).trigger('mouseup') ;
          setTimeout(function(){
              expect(parseInt($('#drag1').offset().left)).to.be(originx + 400) ;
              expect(parseInt($('#drag1').offset().top)).to.be(originy + 600) ;
              done() ;
          }, 80) ;
      }) ;
      
      
      
      // 测试关闭页面拖放
      it('close', function(done){
          var originx = $('#drag1').offset().left ;
              originy = $('#drag1').offset().top ;
              lastx = 0,
              lasty = 0 ;
          
          dnd = new Dnd('#drag1') ;
          $('#drag1').on('mousedown', function(event, x, y){
              event.which = 1 ;
              lastx = event.pageX = originx + x ;
              lasty = event.pageY = originy + y ;
          }) ;
          $('#drag1').on('mousemove', function(event, x, y){
              lastx = event.pageX = lastx + x ;
              lasty = event.pageY = lasty + y ;
          }) ;
          
          dnd.close() ;
          $('#drag1').trigger('mousedown', [20, 20]) ;
          $('#drag1').trigger('mousemove', [200, 300]) ;
          $(document).trigger('mouseup') ;
          setTimeout(function(){
              expect($('#drag1').offset().left).
                      to.be(parseInt(originx)) ;
              expect($('#drag1').offset().top).
                      to.be(parseInt(originy)) ;
              dnd.open() ;
              done() ;
          }, 80) ;
      }) ;
      
      
      
      // 测试边界和方向约束
      it('containment & axis', function(done){
          var originx = $('#drag1').offset().left ;
              originy = $('#drag1').offset().top ;
              lastx = 0,
              lasty = 0 ;
          
          dnd = new Dnd('#drag1', {containment: '#container1', axis: 'x'}) ;
          $('#drag1').on('mousedown', function(event, x, y){
              event.which = 1 ;
              lastx = event.pageX = originx + x ;
              lasty = event.pageY = originy + y ;
          }) ;
          $('#drag1').on('mousemove', function(event, x, y){
              lastx = event.pageX = lastx + x ;
              lasty = event.pageY = lasty + y ;
          }) ;
              
          $('#drag1').trigger('mousedown', [20, 20]) ;
          $('#drag1').trigger('mousemove', [1000, 1000]) ;
          $(document).trigger('mouseup') ;
          setTimeout(function(){
              expect($('#drag1').offset().left + $('#drag1').outerWidth()).to.
                      be($('#container1').offset().left +      
                      $('#container1').outerWidth()) ;
              expect($('#drag1').offset().top).to.
                      be($('#container1').offset().top) ;
              done() ;
          }, 80) ;
      }) ; 
      
      
      
      // 测试drop, 并且没完全置于drop元素中时，将自动将源节点置于drop中央
      it('drop in the center', function(done){
          var originx = $('#drag1').offset().left ;
              originy = $('#drag1').offset().top ;
              lastx = 0,
              lasty = 0 ;
          
          dnd = new Dnd('#drag1', {drop: '#drop1'}) ;
          $('#drag1').on('mousedown', function(event, x, y){
              event.which = 1 ;
              lastx = event.pageX = originx + x ;
              lasty = event.pageY = originy + y ;
          }) ;
          $('#drag1').on('mousemove', function(event, x, y){
              lastx = event.pageX = lastx + x ;
              lasty = event.pageY = lasty + y ;
          }) ;
          
          $('#drag1').trigger('mousedown', [20, 20]) ;
          $('#drag1').trigger('mousemove', [200, 220]) ;
          $(document).trigger('mouseup') ;
          setTimeout(function(){
              expect(parseInt($('#drag1').offset().left)).to.
                      be(parseInt($('#drop1').offset().left + 
                      ($('#drop1').outerWidth() - 
                      $('#drag1').outerWidth()) / 2)) ;
              expect(parseInt($('#drag1').offset().top)).to.
                      be(parseInt($('#drop1').offset().top + 
                      ($('#drop1').outerHeight() - 
                      $('#drag1').outerHeight()) / 2)) ;
              done() ;
          }, 80) ;
      }) ;
      
      
      
      // 测试整个系列事件触发情况以及dataTransfer传输状况
      it('events & dataTransfer', function(done){
          var originx = $('#drag1').offset().left ;
              originy = $('#drag1').offset().top ;
              lastx = 0,
              lasty = 0 ;
          
          dnd = new Dnd('#drag1', {drop: '#drop1'}) ;
          $('#drag1').on('mousedown', function(event, x, y){
              event.which = 1 ;
              lastx = event.pageX = originx + x ;
              lasty = event.pageY = originy + y ;
          }) ;
          $('#drag1').on('mousemove', function(event, x, y){
              lastx = event.pageX = lastx + x ;
              lasty = event.pageY = lasty + y ;
          }) ;
          dnd.on('dragstart', function(dataTransfer, dragging, dropping){
              dataTransfer.data = '我是叁儿' ;
          }) ;
          dnd.on('drop', function(dataTransfer, dragging, dropping){
              dropping.data('message', dataTransfer.data);
          }) ;
          dnd.on('dragenter', function(dragging, dropping){
              dropping.data('isDragenter', true) ;
          }) ;
          dnd.on('dragleave', function(dragging, dropping){
              dropping.data('isDragleave', true) ;
          }) ;
          dnd.on('dragover', function(dragging, dropping){
              dropping.data('isDragover', true) ;
          }) ;
          dnd.on('dragenter', function(dragging, dropping){
              dropping.data('isDragend', true) ;
          }) ;
          
          $('#drag1').trigger('mousedown', [20, 20]) ;
          $('#drag1').trigger('mousemove', [200, 300]) ;
          $('#drag1').trigger('mousemove', [10, 10]) ;
          $('#drag1').trigger('mousemove', [1000, 1000]) ;
          $('#drag1').trigger('mousemove', [-1000, -1000]) ;
          $(document).trigger('mouseup') ;
          setTimeout(function(){
              expect($('#drop1').data('message')).to.be('我是叁儿') ;
              expect($('#drop1').data('isDragenter')).to.be(true) ;
              expect($('#drop1').data('isDragover')).to.be(true) ;
              expect($('#drop1').data('isDragleave')).to.be(true) ;
              expect($('#drop1').data('isDragend')).to.be(true) ;
              done() ;
          }, 80) ;
      }) ;
      
     
     
      // 测试esc将返回源节点初始位置
      it('esc revert', function(done){
          var originx = $('#drag1').offset().left ;
              originy = $('#drag1').offset().top ;
              lastx = 0,
              lasty = 0 ;
          
          dnd = new Dnd('#drag1', {drop:'#drop1', revert: true}) ;
          $('#drag1').on('mousedown', function(event, x, y){
              event.which = 1 ;
              lastx = event.pageX = originx + x ;
              lasty = event.pageY = originy + y ;
          }) ;
          $('#drag1').on('mousemove', function(event, x, y){
              lastx = event.pageX = lastx + x ;
              lasty = event.pageY = lasty + y ;
          }) ;
          $('#drag1').on('keydown', function(event, x, y){
              event.which = 27 ;
          }) ;
          
          $('#drag1').trigger('mousedown', [20, 20]) ;
          $('#drag1').trigger('mousemove', [200, 300]) ;
          $('#drag1').trigger('keydown') ;
          $(document).trigger('mouseup') ;
          setTimeout(function(){
              expect(parseInt($('#drag1').offset().left)).
                      to.be(parseInt(originx)) ;
              expect(parseInt($('#drag1').offset().top)).
                      to.be(parseInt(originy)) ;
              done() ;
          }, 1000) ;
      }) ; 
      
  });

});

src/dnd.js
define(function(require, exports, module){
    var Dnd = null ;
    
    var $ = require('$'),
        Base = require('base') ;
    
    /*
     * static private variable(module global variable)
    */
    var draggingPre = false, // 标识预拖放
        dragging = null, // 当前拖放的代理元素
        dropping = null, // 当前的目标元素
        diffX = 0,
        diffY = 0, // diffX, diffY记录鼠标点击离源节点的距离
        obj = null, // 存储当前拖放的dnd
        dataTransfer = {} ; // 存储拖放信息, 在dragstart可设置,在drop中可读取
        
        

    Dnd = Base.extend({
        attrs: {
            element: {
                value: null,
                readOnly: true
            },
            containment: null,
            axis: false,
            visible: false,
            proxy: null,
            drop: null,
            revert: false,
            revertDuration: 500,
            disabled: false,
            dragCursor: 'move',
            dropCursor: 'copy',
            zIndex: 9999 
        }, 
        
        initialize: function(elem, config){
            var element = null ;
            
            // 检查源节点elem合法性, 初始化dnd
            if($(elem).length === 0 || $(elem).get(0).nodeType !== 1){
                $.error('element error!') ;
            }
            config = $.extend({element: $(elem).eq(0)}, config) ;
            Dnd.superclass.initialize.call(this, config) ;
            element = this.get('element') ;
            
            // 记录下源节点初始style
            if(element.data('style') === undefined){
                if(element.attr('style') === undefined){
                    element.data('style', '') ;
                } else{
                    element.data('style', element.attr('style')) ;
                }
            }
            
            // 在源节点上存储dnd
            element.data('dnd', this) ;
        },
        
        // 拖放期间不能设置配置
        set: function(option, value){
            if(obj !== this){
                Dnd.superclass.set.call(this, option, value, {override: true}) ;
            }
        },
        
        // 开启页面Dnd功能,绑定鼠标按下、移动、释放以及ecs按下事件
        open: function(){
            $(document).on('mousedown', handleEvent) ;
            $(document).on('mousemove', handleEvent) ;
            $(document).on('mouseup', handleEvent) ;
            $(document).on('keydown', handleEvent) ;
        },
        
        // 关闭页面Dnd功能,解绑鼠标按下、移动、释放以及ecs按下事件
        close: function(){
            $(document).off('mousedown', handleEvent) ;
            $(document).off('mousemove', handleEvent) ;
            $(document).off('mouseup', handleEvent) ;
            $(document).off('keydown', handleEvent) ;
        }
    }) ;
    
    
    
    /*
     * 核心部分, 处理鼠标按下、移动、释放以及ecs按下事件, 实现拖放逻辑
    */
    function handleEvent(event){
        var dnd = null ;
        
        switch(event.type){
            case 'mousedown':
                if(event.which === 1){
                    dnd = $(event.target).data('dnd') ;
                    
                    // 判断是否为可拖放元素 用构造函数实例化 或者 
                    // 通过data-dnd=true触发, 此时不支持dataTransfer和一系列事件
                    if(dnd === true){
                        dnd = new Dnd(event.target, 
                              $(event.target).data('config')) ;
                    } else if(dnd instanceof Dnd === true){
                        dnd = $(event.target).data('dnd') ;
                    } else{
                        return ;
                    }
                    
                    // 源节点不允许拖放则返回
                    if(dnd.get('disabled') === true) return ;
                    
                    // 处理配置合法性
                    handleConfig(dnd) ;
                    obj = dnd ;
                    diffX = event.pageX - obj.get('element').offset().left ;
                    diffY = event.pageY - obj.get('element').offset().top ;
                    draggingPre = true ;
                    
                    // 阻止默认选中文本
                    event.preventDefault() ;
                }
                break ;
            
            case 'mousemove':
                if(draggingPre === true){
                    
                    // 开始拖放
                    executeDragStart() ;
                }
                if(dragging !== null){
                    
                    // 根据边界和方向一起判断是否drag并执行
                    executeDrag({pageX: event.pageX, pageY: event.pageY}) ;
                    
                    // 根据dragging和dropping位置来判断
                    // 是否要dragenter, dragleave和dragover并执行
                    executeDragEnterLeaveOver() ;
                    
                    // 阻止默认选中文本
                    event.preventDefault() ;
                }
                break ;
            
            case 'mouseup':
                if(dragging !== null){
                    dragging.css('cursor', 'default') ;
                    dragging.focus() ;
                    dragging = null ;
                    
                    // 根据dropping判断是否drop并执行
                    executeDrop() ;
                    
                    // 根据revert判断是否要返回并执行
                    executeRevert() ;
                    
                    // 此处传递的dragging为源节点element
                    obj.trigger('dragend', obj.get('element'), dropping) ;
                    obj = null ;
                    dropping = null ;
                } else if(draggingPre === true){
                       
                    // 点击而非拖放时
                    obj.get('proxy').remove() ;
                    draggingPre = false ;
                    obj = null ;
                }
                break ;
            
            case 'keydown':
                if(dragging !== null && event.which === 27){
                    dragging.css('cursor', 'default') ;
                    dragging.focus() ;
                    dragging = null ;
                    
                    // 返回源节点初始位置
                    executeRevert(true) ;
                    
                    // 此处传递的dragging为源节点element
                    obj.trigger('dragend', obj.get('element'), dropping) ;
                    obj = null ;
                    dropping = null ;
                }
                break ;
        }
    }
    
    
    
    /*
     * 显示proxy, 按照设置显示或隐藏源节点element
     * 开始拖放
    */
    function executeDragStart(){
        var element = obj.get('element'),
            proxy = obj.get('proxy'),
            visible = obj.get('visible'),
            dragCusor = obj.get('dragCursor'),
            zIndex = obj.get('zIndex') ;
        
        // 按照设置显示或隐藏element
        if(visible === false){
            element.css('visibility', 'hidden') ;
        }
        proxy.css('z-index', zIndex) ;
        proxy.css('visibility', 'visible') ;
        proxy.css('cursor', dragCusor) ;
        proxy.focus() ;
        
        dataTransfer = {} ;
        draggingPre = false ;
        dragging = proxy ;
        obj.trigger('dragstart', dataTransfer, dragging, dropping) ;
    }
    
    
    
    /*
     * 根据边界和方向一起判断是否drag并执行
    */
    function executeDrag(event){
        var container = obj.get('containment'),
            axis = obj.get('axis'),
            xleft = event.pageX - diffX,
            xtop = event.pageY - diffY ;
        
        // 是否在x方向上移动并执行
        if(axis !== 'y'){
            if(container === null || 
                    (xleft >= container.offset().left &&
                    xleft + dragging.outerWidth() <= container.offset().left +
                    container.outerWidth())){
                dragging.css('left', xleft) ;
            } else{
                if(xleft <= container.offset().left){
                    dragging.css('left', container.offset().left) ;
                } else{
                    dragging.css('left', 
                            container.offset().left + container.outerWidth() -  
                            dragging.outerWidth()) ;
                }
            }
        }
        
        // 是否在y方向上移动并执行
        if(axis !== 'x'){
            if(container === null || 
                    (xtop >= container.offset().top &&
                    xtop + dragging.outerHeight() <= container.offset().top +
                    container.outerHeight())){
                dragging.css('top', xtop) ;
            } else{
                if(xtop <= container.offset().top){
                    dragging.css('top', container.offset().top) ;
                } else{
                    dragging.css('top', 
                            container.offset().top + container.outerHeight() -  
                            dragging.outerHeight()) ;
                }
            }
        }
           
        obj.trigger('drag', dragging, dropping) ;
    }
    
    
    
    /*
     * 根据dragging和dropping位置来判断是否dragenter, dragleave和dragover并执行
    */
    function executeDragEnterLeaveOver(){
        var element = obj.get('element'),
            drop = obj.get('drop'),
            dragCursor = obj.get('dragCursor'),
            dropCursor = obj.get('dropCursor'),
            xleft = dragging.offset().left + diffX,
            xtop = dragging.offset().top + diffY ;
        
        if(drop !== null){                    
            if(dropping === null){
                $.each(drop, function(index, elem){
                    
                    // 注意检测drop不是element或者proxy本身
                    if(elem.nodeType === 1 && elem !== element.get(0) &&
                            elem !== dragging.get(0) &&
                            isContain(elem, xleft, xtop) === true){
                        dragging.css('cursor', dropCursor) ;
                        dragging.focus() ;
                        
                        dropping = $(elem) ;
                        obj.trigger('dragenter', dragging, dropping) ;
                        return false ; // 跳出each
                    }
                }) ;
            } else{
                if(isContain(dropping, xleft, xtop) === false){
                    dragging.css('cursor', dragCursor) ;
                    dragging.focus() ;
                    
                    obj.trigger('dragleave', dragging, dropping) ;
                    dropping = null ;
                } else{
                    obj.trigger('dragover', dragging, dropping) ;
                }
            }
        }
    }
    
    
    
    /*
     * 根据dropping判断是否drop并执行
     * 当dragging不在dropping内且不需要revert时, 将dragging置于dropping中央
    */
    function executeDrop(){
        var element = obj.get('element'),
            xdragging = obj.get('proxy'),
            revert = obj.get('revert') ;
        
        if(dropping !== null){
            
            // 放置时不完全在drop中并且不需要返回的放置中央
            if(isContain(dropping, xdragging) === false && revert === false){
                xdragging.css('left', dropping.offset().left +
                        (dropping.outerWidth() -
                        xdragging.outerWidth()) / 2) ;
                xdragging.css('top', dropping.offset().top +
                        (dropping.outerHeight() -
                        xdragging.outerHeight()) / 2) ;
            }
            
            // 此处传递的dragging为源节点element
            obj.trigger('drop', dataTransfer, element, dropping) ;
        }
    }
    
    
    
    /*
     * 根据revert判断是否要返回并执行
     * 若drop不为null且dropping为null, 则自动回到原处
     * flag为true表示必须返回的,目前用于esc按下触发返回
    */
    function executeRevert(flag){
        var element = obj.get('element'),
            xdragging = obj.get('proxy'),
            drop = obj.get('drop'),
            revert = obj.get('revert'),
            revertDuration = obj.get('revertDuration'),
            visible = obj.get('visible'),
            xleft = xdragging.offset().left - element.offset().left,
            xtop = xdragging.offset().top -  element.offset().top ;
            
        if(revert === true || flag === true ||
                (dropping === null && drop !== null)){
            
            //代理元素返回源节点初始位置
            element.attr('style', element.data('style')) ;
            if(visible === false){
                element.css('visibility', 'hidden') ;
            }
            
            xdragging.animate({left: element.offset().left,
                    top: element.offset().top}, revertDuration, function(){
                element.css('visibility', '') ;
                xdragging.remove() ;
            }) ;
        } else{
            
            // 源节点移动到代理元素处
            if(element.css('position') === 'relative'){
                xleft = (isNaN(parseInt(element.css('left'))) ? 0 : 
                        parseInt(element.css('left'))) + xleft ;
                xtop = (isNaN(parseInt(element.css('top'))) ? 0 : 
                        parseInt(element.css('top'))) + xtop ;
            } else if(element.css('position') === 'absolute'){
                xleft = xdragging.offset().left ;
                xtop = xdragging.offset().top ;
            } else{
                element.css('position', 'relative') ;
            }
            
            if(visible === false){
                element.css('left', xleft) ;
                element.css('top', xtop) ;
                element.css('visibility', '') ;
                xdragging.remove() ;
            } else{
                element.animate({left: xleft, top: xtop}, revertDuration, 
                        function(){
                    xdragging.remove() ;
                }) ;  
            }
        }
    }
    
    
    
    /*
     * 检查配置合法性
     * 不合法的配置采用默认配置
     * 每次拖放时都检查一次,防止用户修改配置
    */
    function handleConfig(dnd){
        var element = dnd.get('element'),
            proxy = null,
            flag = false,
            value ;
        
        // containment不能为element本身
        // element也不能在containment外
        value = dnd.get('containment') ;
        if($(value).length === 0 || $(value).get(0).nodeType !== 1 ||
                $(value).get(0) === element.get(0) ||
                isContain($(value).eq(0), element, 1.0) === false){
            dnd.set('containment', null) ;
        } else{
            dnd.set('containment', $(value).eq(0)) ;
        }
        
        // proxy不能为element本身, containment
        // 设置proxy并插入文档, 若在mouseover中插入,会造成抖动
        value = dnd.get('proxy') ;
        if($(value).length === 0 || $(value).get(0).nodeType !== 1 ||
                $(value).get(0) === element.get(0) ||
                $(value).get(0) === $(dnd.get('containment')).get(0)){
            dnd.set('proxy', element.clone()) ;
        } else{
            dnd.set('proxy', $(value).eq(0)) ;
        }
        proxy = dnd.get('proxy') ;
        proxy.css('position', 'absolute') ;
        proxy.css('margin', '0') ;
        proxy.css('left', element.offset().left) ;
        proxy.css('top', element.offset().top) ;
        proxy.css('visibility', 'hidden') ;
        proxy.appendTo('body') ;
        
        // 放置元素不能是elment本身, proxy
        // 若drop中没一个元素符合要求则不合法
        value = dnd.get('drop') ;
        $.each($(value), function(index, elem){
            if(elem.nodeType === 1 && elem !== element.get(0) &&
                    elem !== dnd.get('proxy').get(0)){
                flag = true ;
                return false ; // 跳出each
            }
        }) ;
        if($(value).length === 0 || flag === false){
            dnd.set('drop', null) ;
        } else{
            dnd.set('drop', $(value)) ;
        }
               
        value = dnd.get('axis') ;
        if(value !== 'x' && value !== 'y' && value !== false){
            dnd.set('axis', false) ;
        }       
        value = dnd.get('visible') ;
        if(typeof value !== 'boolean'){
            dnd.set('visible', false) ;
        }       
        value = dnd.get('revert') ;
        if(typeof value !== 'boolean'){
            dnd.set('revert', false) ;
        }       
        value = dnd.get('revertDuration') ;
        if(typeof value !== 'number'){
            dnd.set('revertDuration', 500) ;
        }
        value = dnd.get('disabled') ;
        if(typeof value !== 'boolean'){
            dnd.set('disabled', false) ;
        }
        value = dnd.get('dragCursor') ;
        if(typeof value !== 'string'){
            dnd.set('dragCursor', 'move') ;
        }
        value = dnd.get('dropCursor') ;
        if(typeof value !== 'string'){
            dnd.set('dropCursor', 'copy') ;
        }   
        value = dnd.get('zIndex') ;
        if(typeof value !== 'number'){
            dnd.set('zIndex', 9999) ;
        }
    }
    
    
    
    /*
     * 判断元素B是否位于元素A内部 or 点(B, C)是否位于A内
     * error为了补全IE9,IE10对offset浮点值的差异, 
     * 目前只是在判断container是否合法时使用error=1.0
    */
    function isContain(A, B, C){
        var error = C ;
        
        if(typeof B !== 'number'){
            if(typeof error !== 'number'){
                error = 0 ;
            }
            return $(A).offset().left - error <= $(B).offset().left &&
                    $(A).offset().left + $(A).outerWidth() >= 
                    $(B).offset().left + $(B).outerWidth() - error &&
                    $(A).offset().top - error <= $(B).offset().top && 
                    $(A).offset().top + $(A).outerHeight() >=
                    $(B).offset().top + $(B).outerHeight() - error ;
        } else{  
            return $(A).offset().left <= B &&
                    $(A).offset().left + $(A).outerWidth() >= B &&
                    $(A).offset().top <= C &&
                    $(A).offset().top + $(A).outerHeight() >= C ;
        }
    }
    
    Dnd.prototype.open() ;
    module.exports = Dnd ;
}) ;
